### ISSUE TITLE
Building GTK fails when only GTK3 is installed

### ISSUE THREAD

**Issue by fredo-47** [transmission/transmission#3934]
I am on Linux Mint 21 with GTK3, and was able to successfully build a GTK app from https://github.com/transmission/transmission/commit/aeaa6ca52fcd979a6aaa069b219625c6ba18aed7 a few days ago.

From merged PR https://github.com/transmission/transmission/pull/3918, I understand that building GTK should now be possible with either GTK3 or GTK4, right?

But I have just now tried building GTK from most recent https://github.com/transmission/transmission/commit/0954e287806173438b368b6c7ed634f93e3e612a :

cmake tells me it does not find the files required for GTK4, and apparently ignores that I have GTK3.
```
-- Checking for modules 'gtkmm-4.0>=3.24.0;glibmm-2.68>=2.60.0;giomm-2.68>=2.26.0'
--   No package 'gtkmm-4.0' found
--   No package 'glibmm-2.68' found
--   No package 'giomm-2.68' found
```

And make fails with
```
[ 65%] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/Actions.cc.o
/home/Transmission/gtk/Actions.cc:12:10: fatal error: glibmm.h: No such file or directory
   12 | #include <glibmm.h>
      |          ^~~~~~~~~~
compilation terminated.
make[2]: *** [gtk/CMakeFiles/transmission-gtk.dir/build.make:120: gtk/CMakeFiles/transmission-gtk.dir/Actions.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1349: gtk/CMakeFiles/transmission-gtk.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

How do I tell cmake that I want to build with GTK3, and not GTK4?

**Comment by mikedld** [comment]
Unless you set `USE_GTK_VERSION` option when configuring, CMake will try to look for both GTK 4 and GTK 3 (in that order), using the first one it finds. Seeing the messages you quote is expected in this case; they fail configuration as the build system seems to be generated allowing you to run `make` afterwards.

As for the second error, my guess is that you're missing a development package for GTKMM, which is usually provided separately fro GTK.

**Comment by fredo-47** [comment]
As per your hint, I tried with
`cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo .. -DUSE_GTK_VERSION=3`
That way `make` runs fine and without errors.

It seems, when omitting "USE_GTK_VERSION=3", it does not actually look for GTK3 when it does not find the GTK4.

**Comment by mikedld** [comment]
> It seems, when omitting "USE_GTK_VERSION=3", it does not actually look for GTK3 when it does not find the GTK4.

From the [CI build](https://github.com/transmission/transmission/actions/runs/3212435842/jobs/5251305643) for the PR you mention, which doesn't set `USE_GTK_VERSION`:
```
  cmake \
    -S src \
    -B obj \
    -G Ninja \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DCMAKE_INSTALL_PREFIX=pfx \
    -DENABLE_CLI=ON \
    -DENABLE_DAEMON=ON \
    -DENABLE_GTK=ON \
    -DENABLE_MAC=OFF \
    -DENABLE_QT=ON \
    -DENABLE_TESTS=ON \
    -DENABLE_UTILS=ON \
    -DENABLE_WEB=OFF \
    -DRUN_CLANG_TIDY=OFF
```
it searches for both versions:
```
-- Checking for modules 'gtkmm-4.0>=3.24.0;glibmm-2.68>=2.60.0;giomm-2.68>=2.26.0'
--   No package 'gtkmm-4.0' found
--   No package 'glibmm-2.68' found
--   No package 'giomm-2.68' found
-- Checking for modules 'gtkmm-3.0>=3.24.0;glibmm-2.4>=2.60.0;giomm-2.4>=2.26.0'
--   Found gtkmm-3.0, version 3.24.2
--   Found glibmm-2.4, version 2.64.2
--   Found giomm-2.4, version 2.64.2
-- Checking for module 'appindicator3-0.1>=0.4.90'
--   No package 'appindicator3-0.1' found
```
and then builds GTK client just fine:
```
...
[244/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/Actions.cc.o
...
[247/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/Dialogs.cc.o
[248/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/FaviconCache.cc.o
[249/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/Application.cc.o
[250/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/DetailsDialog.cc.o
[251/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/FileList.cc.o
[252/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/FreeSpaceLabel.cc.o
[253/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/IconCache.cc.o
[254/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/main.cc.o
[255/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/FilterBar.cc.o
[256/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/MakeDialog.cc.o
[257/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/MainWindow.cc.o
[258/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/MessageLogWindow.cc.o
[259/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/Notify.cc.o
[260/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/OptionsDialog.cc.o
[261/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/PathButton.cc.o
[262/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/Prefs.cc.o
[263/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/RelocateDialog.cc.o
[264/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/PrefsDialog.cc.o
[265/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/Session.cc.o
[266/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/StatsDialog.cc.o
[267/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/SystemTrayIcon.cc.o
[268/375] Building C object gtk/CMakeFiles/transmission-gtk.dir/transmission-resources.c.o
[269/375] Building C object gtk/CMakeFiles/transmission-gtk.dir/transmission-ui-resources.c.o
[270/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/TorrentCellRenderer.cc.o
...
[315/375] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/Utils.cc.o
[316/375] Linking CXX executable gtk/transmission-gtk
...
```

**Comment by mikedld** [comment]
I know what's going on but I don't think I want to fix it.

The reason you're having an issue is because previously GTK search results were written into variables prefixed `GTK_`, including `GTK_FOUND`; after my changes, they're now written in two sets of variables prefixed `GTK4_` (+`GTK4_FOUND`) and `GTK3_` (+`GTK3_FOUND`), and the combined result is stored in a variable called `GTK_FOUND`. Since you were using the same build directory before, and since `GTK_FOUND` variable was cached back then, you now have `GTK_FOUND` set to `TRUE` (loaded from CMakeCache.txt) which leads to skipping search for GTK 3 while leaving a whole set of variables prefixed `GTK3_` undefined. It would've worked if only we weren't trying to find GTK 4 and in the process changing `GTK_VERSION` from empty value to `4`, which then leads to further GTK client configuration trying to use variables prefixed `GTK4_` which aren't defined either.

In other words, clean rebuild should help you.

**Comment by fredo-47** [comment]
I see, that's it. Thank you for the detailed explanation!
Indeed, I emptied my build directory and now it works again like a charm :-)
